Clover/app/src/main/java/org/floens/chan/core/manager/PageRequestManager.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Clover - 4chan browser https://github.com/Floens/Clover/
+ * Copyright (C) 2014  Floens
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.floens.chan.core.manager;
+
+import org.floens.chan.core.model.Post;
+import org.floens.chan.core.model.orm.Board;
+import org.floens.chan.core.model.orm.Loadable;
+import org.floens.chan.core.site.Page;
+import org.floens.chan.core.site.Pages;
+import org.floens.chan.core.site.SiteActions;
+import org.floens.chan.core.site.ThreadTime;
+import org.floens.chan.utils.Logger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class PageRequestManager implements SiteActions.PagesListener {
+    private static final String TAG = "PageRequestManager";
+
+    private Set<String> requestedBoards = Collections.synchronizedSet(new HashSet<>());
+    private Set<String> savedBoards = Collections.synchronizedSet(new HashSet<>());
+    private ConcurrentMap<String, Pages> boardPagesMap = new ConcurrentHashMap<>();
+    private ConcurrentMap<String, Long> boardTimeMap = new ConcurrentHashMap<>();
+
+    private List<PageCallback> callbackList = new ArrayList<>();
+
+    @Inject
+    public PageRequestManager() {
+    }
+
+    public Page getPage(Post op) {
+        if (op == null) {
+            return null;
+        }
+        Pages pages = getPages(op.board);
+        if (pages == null) {
```

Comment:
could we dedupe the code here?

maybe it could be a method on the `Pages` model? `pages.findPageForThread(ThreadTime threadTime)` to hide this logic away

Reply:
Sure, the logic is basically the same.